### DISCUSSION TITLE
Trigger field array level validation using field values

### DISCUSSION THREAD

**Discussion by LauSam09** [react-hook-form/react-hook-form#2960] [Q&A]
Apologies if this has already been answered - I've followed the docs for `useFieldArray`, specifically this section
```js
// trigger validation at the field array level
useEffect(() => {
  //if (fields) trigger('test'); // with resolver
  if (!fields.length) setError('miniLength, 'at least 1 field array item'); // build-in validator
}, [trigger, fields])
```

and this works great for `append`, `remove`, etc., but I am trying to check for duplicate values across the array, so would like to know the recommended way to fire validation (and retrieve values) when any of the fields in the array change.

This is what I have so far, but it isn't executing if I update values in the array, only if I modify the array itself..
```js
const { fields, append } = useFieldArray({
  control,
  name: 'settings'
});

useEffect(() => {
  if (fields) {
    const uniqueValues = new Set(fields.map(s => s.angle));
    if (uniqueValues.size < fields.length) {
      setError('settingsArray', { type: 'duplicate', message: 'Duplicate angles' });
    }
  }
}, [fields, setError]);
```

I've looked at using `trigger` in the `onChange` event, and `watch` on the array, but haven't had much luck. 

Thanks in advance!

**Comment by bluebill1049** [comment]
for array field validation, what you got is correct, this is one of the strengh with schema validation or resolver, because the validation is at the form level, while build-in validation is at the field level.

**Reply by LauSam09** [reply]
Thanks for your answer. I started writing a custom resolver to add validation for the array, but found that it replaces all of the build-in validation that I've got (and I don't really want to move it - there's a _lot_).. I'm surprised that operations over an array aren't a more common use-case.

The only workaround I have is to add the validation to the individual fields and trigger validation on all of them when any change...
![image](https://user-images.githubusercontent.com/24778857/93904516-d808e180-fcf1-11ea-90a5-a4d489327f5e.png)

But this isn't great performance-wise (and with my lazy coding duplicates the error for every field in the array).

In case you have any suggestions I'll leave this open a while before marking as answered. Thanks in advance and great library!

**Reply by bluebill1049** [reply]
yes agree above, i would try to trigger only at useEffect (maybe)?

**Reply by LauSam09** [reply]
Then I'm back to my original issue of only getting my validation when I add/remove an element in the array.
If I have 3 fields in the array:
```
[ 10 ]
[ 20 ]
[ 30 ]
```
and I change the values to be duplicates:
```
[ 10 ]
[ 10 ]
[ 10 ]
```
then the effect never fires as the `fields` object has not changed, and I get no validation message.

**Reply by vatsalsood** [reply]
I am facing the same issue. Were you able to find an efficient solution to this @LauSam09  ?

**Reply by LauSam09** [reply]
@vatsalsood I'm afraid I wasn't able to - I can't remember exactly what we did but I suspect we didn't use react-hook-form for this kind of validation.

**Reply by vatsalsood** [reply]
Thanks for letting me know @LauSam09. I achieved what I wanted to do with a combination of watch and the onChange event handler. I had to account for a lot of different scenarios though with some loops etc.

**Reply by vrato-sabo** [reply]
@vatsalsood could you post a code of how you achieved it?
